devtools/server/actors/replay/module.js

```diff
@@ -770,6 +771,10 @@ function consoleAPIMessageLevel({ level }) {
       return "error";
     case "assert":
       return "assert";
+     case "group":
+      return "group";
+    case "groupEnd":
+      return "groupEnd";
```

Comment:
"groupCollapsed" ? 😁 

😮 "table" ???

Reply:
good call. Here are more log levels from firefox
https://searchfox.org/mozilla-central/source/toolkit/modules/Console.jsm#285-304
https://searchfox.org/mozilla-central/source/devtools/client/webconsole/utils/messages.js#563-583

Reply:
RE: Nesting:

I _think_ it's probably reasonable to pass the potentially unmatched groups to the frontend and rely on it to do a smart thing (like ignoring group-close entries without group-opens, or auto-closing unclosed groups).

Reply:
Yes, maybe! Although I think it depends on what we mean by "consistently"

Reply:
>  it really feels like we'd need console messages themselves to individually be aware of their depth for things to work consistently

Maybe! I thought about this a little bit though and I think it depends on what decide is the desired behavior. 

I think there are three scenarios, and there's only one that we can't handle "right" (depending on what "right" is).

# Scenario 1: Implicitly closed group(s)
```js
// Focus window starts here
console.group("This is a group");
console.log("This is a log");
// Focus window ends here
console.groupEnd();
```
We can auto-close any lingering groups. (We'll need to do this anyway because of overflow.) This case shouldn't matter at all.

# Scenario 2: Implicitly opened group(s)
```js
console.group("This is a group");
// Focus window starts here
console.log("This is a log");
console.groupEnd();
// Focus window ends here
```
We can reliably infer the existence of a `group` (or groups) from the presence of a `groupEnd`, but we can't distinguish between `group` and `groupCollapsed` so there's some chance we'd do the "wrong thing" here.

# Scenario 3: Ambiguous
```js
console.group("This is a group");
// Focus window starts here
console.log("This is a log");
// Focus window ends here
console.groupEnd();
```
We can't infer anything here, so we'd show the. log as though it's not grouped at all (no indent). Maybe this is bad or maybe it's fine? Groups exist to visually separate logs, and if there's nothing to separate them "from" then maybe they aren't necessary?

Reply:
I hadn't even considered the behavior in the context of windows, but I guess to me it seems like we'd still want to preserve the the nesting? Like if I have
```
console.group("Beginning App Thing");
// Focus window starts here
console.group("Beginning sub-task");
console.group("Beginning sub-sub-task");
console.log("This is a first log");
console.groupEnd();
console.log("This is a second log");
console.groupEnd();
console.log("This is a third log");
console.groupEnd();
console.log("This is a fourth log");
// Focus window ends here
```
like we'd still ideally want to show the logs each at their own depth. If we just start with the message at the start of the window, indent twice, and then dedent on group end, the third and fourth log will end up at the same level.

> We can auto-close any lingering groups. (We'll need to do this anyway because of overflow.)

Overflow is really the main thing that was driving this for me, so if we had like
```
// Focus window starts here
// start initial region
console.group("Beginning App Thing"); // 0 ->1 depth
console.group("Quick loop of stuff"); // 1->2 depth
for (let i = 0; i < 999; i++) {
  console.log("value1", i); // 2 depth
} 
// this will be ignored and overflow:true indicated
console.groupEnd(); // 2->1 depth

// new region triggered somehow so overflow reset

for (let i = 0; i < 1000; i++) {
  console.log("value2", i); // 1 depth expected
}
// this will be ignored and overflow:true indicated
console.group("another group"); // 1->2 depth

// new region starts, clearing overflow

for (let i = 0; i < 1200; i++) {
  console.log("value3", i); // 2 depth
}
console.groupEnd(); // 2->1 depth

console.log("other thing"); // 1 depth
// Focus window ends here
```
the first group is meant to apply to _all_ of these messages, so if the overflow at the first `groupEnd` cleared all groups, all later messages would be nested at the wrong level. Similarly if the first groupEnd is lost in the overflow and we _don't_ pop everything, then the `value2` logs would log with depth 2 instead of depth one, and there's similar weirdness for all the `value3` logs.

Reply:
It's a little unclear to me if you are describing a case/scenario that I didn't list above?

Reply:
Fair, should have been clearer. It seems like the randomness of recordings and of recording region division mean that any particular log statement may end up falling into your "Ambiguous" item.

What I'm trying to illustrate with my examples is that I don't think we _can_ fix the ambiguous cases, and I'm realizing a detail that I should have clarified, which is that the linker itself is what tracks `overflow`, so when there are overflowing messages, we literally know nothing, even in the backend, about those messages, for

> We could consider having the backend always append any group/groupCollapsed entries to the start of a log chunk so that we can avoid scenarios 2 and 3 above, but I don't know how messy that might be on the backend. Also happy to talk about alternate approaches!

We do not have enough information in the backend to know how to do this for the ambiguous case, because we would need to know when groups may have been opened or closed in the overflowed messages.

So either the messages themselves need to know their depth, or the linker itself needs to track message depth for overflowing messages, even if it discards the message itself.

Reply:
Gotcha. Thank you for distilling that. 

> So either the messages themselves need to know their depth

I'm not sure what this means. Are you saying that the gecko/chromium/node would have to store this information at runtime/record-time?

I have no idea how much effort that would be. If it's easy to do, then– that would be a great option to have.

I think we might still be skipping past what the _desired behavior_ is for the scenarios I raised above. Like maybe the thread moved forward under the assumption that preserving the indentation was definitely what we wanted to do– but I'm not sure it is.

For instance, what Chrome does when you clear the console feels somewhat analogous to how our focus window feature works? In this case, Chrome blows away the indentation:
![Kapture 2022-07-16 at 07 33 19](https://user-images.githubusercontent.com/29597/179353120-077c6ff8-3577-49a7-b8eb-e8a469ca78d7.gif)